File: docs/js/geometry.js
// Geometry: 3x3 rotations (row-major Float64Array(9)), the fisheye lens model
// and the equirectangular conventions.  Mirrors insta360stitch/geometry.py.
export const DEG = Math.PI / 180;

export function mat3Identity() { return new Float64Array([1, 0, 0, 0, 1, 0, 0, 0, 1]); }
export function mat3Mul(a, b) {
  const o = new Float64Array(9);
  for (let i = 0; i < 3; i++) for (let j = 0; j < 3; j++)
    o[i * 3 + j] = a[i * 3] * b[j] + a[i * 3 + 1] * b[3 + j] + a[i * 3 + 2] * b[6 + j];
  return o;
}
export function mat3T(a) { return new Float64Array([a[0], a[3], a[6], a[1], a[4], a[7], a[2], a[5], a[8]]); }
export function mat3Apply(m, v) {
  return [m[0] * v[0] + m[1] * v[1] + m[2] * v[2], m[3] * v[0] + m[4] * v[1] + m[5] * v[2], m[6] * v[0] + m[7] * v[1] + m[8] * v[2]];
}
export function rotX(deg) { const c = Math.cos(deg * DEG), s = Math.sin(deg * DEG); return new Float64Array([1, 0, 0, 0, c, -s, 0, s, c]); }
export function rotY(deg) { const c = Math.cos(deg * DEG), s = Math.sin(deg * DEG); return new Float64Array([c, 0, s, 0, 1, 0, -s, 0, c]); }
export function rotZ(deg) { const c = Math.cos(deg * DEG), s = Math.sin(deg * DEG); return new Float64Array([c, -s, 0, s, c, 0, 0, 0, 1]); }
export function eulerYPR(yaw, pitch, roll) { return mat3Mul(mat3Mul(rotY(yaw), rotX(pitch)), rotZ(roll)); }
export function normalize(v) { const n = Math.hypot(v[0], v[1], v[2]) || 1; return [v[0] / n, v[1] / n, v[2] / n]; }
function cross(a, b) { return [a[1] * b[2] - a[2] * b[1], a[2] * b[0] - a[0] * b[2], a[0] * b[1] - a[1] * b[0]]; }

/** Minimal rotation R with R a = b (unit vectors). */
export function rotationBetween(a, b) {
  a = normalize(a); b = normalize(b);
  const v = cross(a, b);
  const c = a[0] * b[0] + a[1] * b[1] + a[2] * b[2];
  const s = Math.hypot(v[0], v[1], v[2]);
  if (s < 1e-12) {
    if (c > 0) return mat3Identity();
    let axis = cross(a, [1, 0, 0]);
    if (Math.hypot(...axis) < 1e-6) axis = cross(a, [0, 1, 0]);
    axis = normalize(axis);
    const o = new Float64Array(9);
    for (let i = 0; i < 3; i++) for (let j = 0; j < 3; j++) o[i * 3 + j] = 2 * axis[i] * axis[j] - (i === j ? 1 : 0);
    return o;
  }
  const k = [v[0] / s, v[1] / s, v[2] / s];
  const K = new Float64Array([0, -k[2], k[1], k[2], 0, -k[0], -k[1], k[0], 0]);
  const KK = mat3Mul(K, K);
  const o = mat3Identity();
  for (let i = 0; i < 9; i++) o[i] += s * K[i] + (1 - c) * KK[i];
  return o;
}
/** World rotation R with R (0,1,0) = upCam. */
export function levelRotation(upCam) { return rotationBetween([0, 1, 0], upCam); }
/** ONE X accelerometer (sensor frame) -> camera-frame up vector in our world axes. */
export function upFromAccel(a) { return normalize([a[2], -a[0], a[1]]); }

export function equirectDir(u, v, W, H) {
  const lon = ((u + 0.5) / W * 2 - 1) * Math.PI;
  const lat = (0.5 - (v + 0.5) / H) * Math.PI;
  const cl = Math.cos(lat);
  return [cl * Math.sin(lon), Math.sin(lat), cl * Math.cos(lon)];
}
export function lonToU(lonDeg, W) { return (lonDeg / 180 + 1) * 0.5 * W; }

export class Lens {
  constructor(o) {
    Object.assign(this, { cx: 0, cy: 0, rCal: 1, fov: 200, k1: 0, k2: 0, yaw: 0, pitch: 0, roll: 0, flip: false, edgeFrac: 0.985 }, o);
  }
  clone() { return new Lens(this); }
  g(t) { const t2 = t * t; return t * (1 + t2 * (this.k1 + t2 * this.k2)); }
  dg(t) { const t2 = t * t; return 1 + t2 * (3 * this.k1 + 5 * this.k2 * t2); }
  get f() { return this.rCal / this.g(this.fov * 0.5 * DEG); }
  thetaOfR(r) { const x = r / this.f; let t = x; for (let i = 0; i < 6; i++) t -= (this.g(t) - x) / this.dg(t); return t; }
  get thetaMax() { return this.thetaOfR(this.edgeFrac * this.rCal); }
  /** world -> lens rotation */
  M() { let R = eulerYPR(this.yaw, this.pitch, this.roll); if (this.flip) R = mat3Mul(R, rotY(180)); return mat3T(R); }
  /** lens-frame unit vector -> [X, Y, theta] in the stored image */
  project(v) {
    const theta = Math.acos(Math.max(-1, Math.min(1, v[2])));
    const rho = Math.max(Math.hypot(v[0], v[1]), 1e-9);
    const r = this.f * this.g(theta);
    return [this.cx + r * v[1] / rho, this.cy + r * v[0] / rho, theta];
  }
  /** stored image pixel -> lens-frame unit vector */
  unproject(X, Y) {
    const dx = X - this.cx, dy = Y - this.cy;
    const r = Math.hypot(dx, dy);
    const theta = this.thetaOfR(r);
    const s = Math.sin(theta) / Math.max(r, 1e-9);
    return [dy * s, dx * s, Math.cos(theta)];
  }
  toJSON() { const o = {}; for (const k of ["cx", "cy", "rCal", "fov", "k1", "k2", "yaw", "pitch", "roll", "flip", "edgeFrac"]) o[k] = this[k]; return o; }
}

/** Build the two lenses from a parsed calibration (see insp.js). */
export function lensesFromCalibration(cal, imgW, imgH, fov = 200) {
  const centres = cal.storedCentres(imgW, imgH);
  return cal.lenses.map((lc, i) => new Lens({ cx: centres[i][0], cy: centres[i][1], rCal: centres[i][2], fov, yaw: lc.yaw, pitch: lc.pitch, roll: lc.roll, flip: i === 1 }));
}

File: insta360stitch/geometry.py
"""Camera geometry: rotations, equirectangular grids and the fisheye lens model.

World frame: x right, y up, z forward (the optical axis of lens 1).
Equirectangular output: longitude -180..180 left to right (0 at the centre),
latitude +90 at the top.
"""
from __future__ import annotations

import math
from dataclasses import dataclass

import cv2
import numpy as np

DEG = math.pi / 180.0


def rot_x(deg: float) -> np.ndarray:
    a = deg * DEG
    c, s = math.cos(a), math.sin(a)
    return np.array([[1, 0, 0], [0, c, -s], [0, s, c]], dtype=np.float64)


def rot_y(deg: float) -> np.ndarray:
    a = deg * DEG
    c, s = math.cos(a), math.sin(a)
    return np.array([[c, 0, s], [0, 1, 0], [-s, 0, c]], dtype=np.float64)


def rot_z(deg: float) -> np.ndarray:
    a = deg * DEG
    c, s = math.cos(a), math.sin(a)
    return np.array([[c, -s, 0], [s, c, 0], [0, 0, 1]], dtype=np.float64)


def euler_ypr(yaw: float, pitch: float, roll: float) -> np.ndarray:
    return rot_y(yaw) @ rot_x(pitch) @ rot_z(roll)


def rotation_between(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    """Minimal rotation R with R @ a = b for unit vectors a, b."""
    a = np.asarray(a, float) / np.linalg.norm(a)
    b = np.asarray(b, float) / np.linalg.norm(b)
    v = np.cross(a, b)
    c = float(np.dot(a, b))
    s = float(np.linalg.norm(v))
    if s < 1e-12:
        if c > 0:
            return np.eye(3)
        # 180 degrees: pick any axis orthogonal to a
        axis = np.cross(a, [1, 0, 0])
        if np.linalg.norm(axis) < 1e-6:
            axis = np.cross(a, [0, 1, 0])
        axis /= np.linalg.norm(axis)
        return 2 * np.outer(axis, axis) - np.eye(3)
    k = v / s
    K = np.array([[0, -k[2], k[1]], [k[2], 0, -k[0]], [-k[1], k[0], 0]])
    return np.eye(3) + s * K + (1 - c) * (K @ K)


def equirect_dirs(width: int, height: int, y0: int = 0, y1: int | None = None, lon_offset_deg: float = 0.0) -> np.ndarray:
    """Unit direction vectors for equirect pixels rows y0..y1, shape (rows, width, 3)."""
    if y1 is None:
        y1 = height
    lon = ((np.arange(width, dtype=np.float32) + 0.5) / width * 2.0 - 1.0) * np.pi + lon_offset_deg * DEG
    lat = (0.5 - (np.arange(y0, y1, dtype=np.float32) + 0.5) / height) * np.pi
    cl = np.cos(lat)[:, None]
    sl = np.sin(lat)[:, None]
    d = np.empty((y1 - y0, width, 3), np.float32)
    d[..., 0] = cl * np.sin(lon)[None, :]
    d[..., 1] = np.broadcast_to(sl, (y1 - y0, width))
    d[..., 2] = cl * np.cos(lon)[None, :]
    return d


def dirs_to_equirect(d: np.ndarray, width: int, height: int):
    """Inverse of equirect_dirs: directions (...,3) -> pixel (u, v) float arrays."""
    lon = np.arctan2(d[..., 0], d[..., 2])
    lat = np.arcsin(np.clip(d[..., 1], -1, 1))
    u = (lon / np.pi + 1.0) * 0.5 * width - 0.5
    v = (0.5 - lat / np.pi) * height - 0.5
    return u, v


def lon_to_u(lon_deg: float, width: int) -> float:
    return (lon_deg / 180.0 + 1.0) * 0.5 * width


@dataclass
class Lens:
    """Fisheye lens model in the stored JPEG frame.

    Radial model r(theta) = f * (theta + k1 theta^3 + k2 theta^5) with f
    chosen so that r(fov/2) == r_cal, i.e. `fov` is the field of view at the
    calibrated circle radius.  The image-plane orientation of the lens is
    such that stored X grows with lens-frame y and stored Y with lens-frame
    x (this is how the ONE X stores its sensor, rotated by 90 degrees).

    `rotation` is (yaw, pitch, roll) in degrees; `flip` is True for the rear
    lens which looks along -z.  M() returns the world->lens rotation.
    """

    cx: float
    cy: float
    r_cal: float
    fov: float = 200.0
    k1: float = 0.0
    k2: float = 0.0
    yaw: float = 0.0
    pitch: float = 0.0
    roll: float = 0.0
    flip: bool = False
    edge_frac: float = 0.985

    # ---- radial model -------------------------------------------------
    def _g(self, theta):
        t2 = theta * theta
        return theta * (1.0 + t2 * (self.k1 + t2 * self.k2))

    def _dg(self, theta):
        t2 = theta * theta
        return 1.0 + t2 * (3.0 * self.k1 + 5.0 * self.k2 * t2)

    @property
    def f(self) -> float:
        return self.r_cal / float(self._g(self.fov * 0.5 * DEG))

    def theta_of_r(self, r):
        """Invert the radial model (Newton), r in pixels -> theta in radians."""
        x = np.asarray(r, np.float64) / self.f
        theta = x.copy()
        for _ in range(6):
            theta = theta - (self._g(theta) - x) / self._dg(theta)
        return theta

    @property
    def theta_max(self) -> float:
        return float(self.theta_of_r(self.edge_frac * self.r_cal))

    def M(self) -> np.ndarray:
        R = euler_ypr(self.yaw, self.pitch, self.roll)
        if self.flip:
            R = R @ rot_y(180.0)
        return R.T

    # ---- projection ---------------------------------------------------
    def project(self, v: np.ndarray):
        """Lens-frame unit vectors (...,3) -> (X, Y, theta) in the stored image."""
        vx, vy, vz = v[..., 0], v[..., 1], v[..., 2]
        theta = np.arccos(np.clip(vz, -1.0, 1.0))
        rho = np.hypot(vx, vy)
        rho = np.where(rho < 1e-9, 1e-9, rho)
        r = self.f * self._g(theta)
        X = self.cx + r * vy / rho
        Y = self.cy + r * vx / rho
        return X, Y, theta

    def unproject(self, X, Y) -> np.ndarray:
        """Stored image pixels -> lens-frame unit vectors (...,3)."""
        dx = np.asarray(X, np.float64) - self.cx
        dy = np.asarray(Y, np.float64) - self.cy
        r = np.hypot(dx, dy)
        theta = self.theta_of_r(r)
        s = np.sin(theta) / np.where(r < 1e-9, 1e-9, r)
        v = np.stack([dy * s, dx * s, np.cos(theta)], axis=-1)
        return v

    def to_dict(self) -> dict:
        return {k: getattr(self, k) for k in ("cx", "cy", "r_cal", "fov", "k1", "k2", "yaw", "pitch", "roll", "flip", "edge_frac")}

    @classmethod
    def from_dict(cls, d: dict) -> "Lens":
        return cls(**{k: d[k] for k in ("cx", "cy", "r_cal", "fov", "k1", "k2", "yaw", "pitch", "roll", "flip", "edge_frac") if k in d})


def lenses_from_calibration(cal, img_w: int, img_h: int, fov: float = 200.0):
    """Build the two Lens objects from a parsed factory calibration."""
    centres = cal.stored_centres(img_w, img_h)
    lenses = []
    for i, (lc, (X, Y, r)) in enumerate(zip(cal.lenses, centres)):
        lenses.append(Lens(cx=X, cy=Y, r_cal=r, fov=fov, yaw=lc.yaw, pitch=lc.pitch, roll=lc.roll, flip=(i == 1)))
    return lenses


def render_lens(img: np.ndarray, lens: Lens, R_world: np.ndarray, width: int, height: int,
                chunk: int = 256, u_range=None):
    """Remap one fisheye image to equirectangular.

    R_world is an extra world rotation (levelling); the total world->lens
    rotation is lens.M() @ R_world.  Returns (image, valid_mask).
    u_range (u0, u1) limits the rendered columns; the returned arrays then
    have width u1-u0.
    """
    u0, u1 = (0, width) if u_range is None else u_range
    W = u1 - u0
    out = np.zeros((height, W, 3), np.uint8)
    valid = np.zeros((height, W), bool)
    RT = (lens.M() @ R_world).T.astype(np.float32)
    theta_max = lens.theta_max
    h_img, w_img = img.shape[:2]
    margin = 1.5
    for y0 in range(0, height, chunk):
        y1 = min(height, y0 + chunk)
        d = equirect_dirs(width, height, y0, y1)[:, u0:u1]
        v = d @ RT
        X, Y, theta = lens.project(v)
        ok = (theta < theta_max) & (X > margin) & (X < w_img - 1 - margin) & (Y > margin) & (Y < h_img - 1 - margin)
        mapx = np.where(ok, X, -1e4).astype(np.float32)
        mapy = np.where(ok, Y, -1e4).astype(np.float32)
        out[y0:y1] = cv2.remap(img, mapx, mapy, cv2.INTER_LINEAR, borderMode=cv2.BORDER_CONSTANT)
        valid[y0:y1] = ok
    return out, valid


def level_rotation(up_cam: np.ndarray) -> np.ndarray:
    """World rotation R such that camera-frame `up_cam` becomes world +y.

    render_lens applies v = M @ R @ d, so we need R with R @ (0,1,0) = up_cam.
    """
    return rotation_between(np.array([0.0, 1.0, 0.0]), up_cam)
